Sources/LanguageServerProtocol/Command.swift

```diff
@@ -0,0 +1,133 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SKSupport
+
+/// A `CommandDataType` represents the underlying data of a custom `Command`.
+public protocol CommandDataType: Codable {
+  static var identifier: String { get }
+  var title: String { get set }
+  var textDocument: TextDocumentIdentifier { get set }
+}
+
+extension CommandDataType {
+
+  fileprivate typealias EncodingContainer = KeyedEncodingContainer<Command.CodingKeys>
+  fileprivate typealias DecodingContainer = KeyedDecodingContainer<Command.CodingKeys>
+
+  fileprivate static func decodeLSPArgumentKey(fromContainer container: DecodingContainer) throws -> Self {
+    let arguments: [Self] = try container.decode([Self].self, forKey: .arguments)
+    guard arguments.count == 1 else {
+      log("Tried to decode command with an invalid argument structure", level: .warning)
+      throw Command.CodingError.invalidArguments
+    }
+    return arguments[0]
+  }
+
+  fileprivate func encodeForLSPArgumentKey(inContainer container: inout EncodingContainer) throws {
+    try container.encode([self], forKey: .arguments)
+  }
+}
+
+/// Represents a reference to a command identified by a string. Used as the result of
+/// requests that returns actions to the user, later used as the parameter of
+/// workspace/executeCommand if the user wishes to execute said command.
+public struct Command {
+
+  /// Title of the command, like `save`.
+  public var title: String {
+    return data.title
+  }
+
+  /// The internal identifier for this command.
+  public var command: String {
+    return type(of: data).identifier
+  }
+
+  /// The data related to this command.
+  /// This is called `arguments` and is [Any]? in the LSP, but internally we treat it
+  /// differently to make it easier to create and (de)serialize commands.
+  public var data: CommandDataType
```

Comment:
Also, could we keep the name `arguments` for this field, just like in the LSP spec?

Reply:
Thanks for the comments Daniel! I was concerned that directly naming it `arguments` would be misleading since it works differently from the specs, but I agree that it sounds better.

I made it non-optional because the command identifier comes from the protocol, but I don't remember why I did that. I think making it optional will be better indeed 👍

Reply:
I've been applying the suggestions and I now remember why I did that - it's because SourceKit-LSP requires all requests to have a `textDocument` and ExecuteCommand isn't a `TextDocumentRequest`. I forced the argument to always contain it in order to circumvent it because I assumed that making the requests not need it would be a massive change, if possible at all since it uses that to determine which language server to trigger. I'm not sure what's the best approach here now. What do you think?

Reply:
Can you expand on this? We should not require a textDocument in general.  If the Command is something that will be executed by sourcekitd/clangd then sure, we need to know how to dispatch it, but that doesn't need to be a document name.  We can embed whatever information (e.g. toolchain service) we need into `arguments` and then look it up if we are asked to execute the command.

Reply:
Ah, I was under the impression that we couldn't support argument-less commands in the code as we somehow have to explicitly add the "this is Swift" info in the command's arguments (and a document was the easier way since the requests are built on it), but after some thinking I reworked the structure to make Command's `Codable` inject/retrieve the `textDocument` in/from the first position of the array automatically. The command's JSON itself will never have empty arguments, but the type inside the code will now be able to. 

This is going to do the trick while we don't have a command without a document related to it, but I see now that adding a type of request that doesn't need a document would be straight-forward if we end up getting to that point.